src/types/common.d.ts

```diff
@@ -109,3 +109,25 @@ export interface ModifierContextState {
   text_line: number;
   top_text_line: number;
 }
+
+
+/* TODO: Move to EasyScore / Grammar / Parser */
+export type Match = string | Match[] | null;
```

Comment:
This can be in `parser.ts`

Reply:
@0xfe I also commented on the topic that we are moving from `null` to `undefined`. Can you comment on that? @ronyeh 's concern is the change on the interface. Can you comment on that? That was already discussed in the past.

Reply:
Good question. In general, `undefined` for uninitialized values, and `null` when we specifically mean "none, empty set, not found, etc." -- so here, `null` is good.

Reply:
This is a normal approach agree. But I believe we discussed that we wanted to use either `undefined` or `null` and moved to all using `undefined` in order not to have to include `null` in types like the example above and in order to use ? in optional attributes and params as @wassertim suggested. This is just a convention, which is also normal, but if we change it now we will be inconsistent with the rest. We decided to move to:

```
class A {
  reqString: string;
  optString?: string;
};

functionA(  reqString: string; optString?: string;): string | undefined
```
and not use:
```
class A {
  reqString: string;
  optString: string | null;
};

functionA(  reqString: string; optString: string | null;): string | null
```
I started with `null` and I changed it. :)

Reply:
Right, your example is fine because semantically `optString` can be defined or undefined. That is different from `null`, which, semantically, means defined and has some kind of value. It's like `undefined` vs `0` for `number`, or `undefined` vs `""` for `string`.

In this code `Match` as `null` means no match, as opposed to (not defined, don't know, etc.) The way I see it, `null` should only be used when it means something specific (typically [zero value](https://dave.cheney.net/2013/01/19/what-is-the-zero-value-and-why-is-it-useful)) -- `undefined` when it doesn't have a meaning, or the meaning is not known.

Reply:
@rvilarl I mostly disagree with that Internet comment -- I think it completely disregards the nuances :-)

> Null and undefined are essentially two different values that mean the same thing.

Null and undefined are different zero values with different semantics. The former is an explicit semantic, the latter is not. See my example below.

> The only difference is in the conventions of how you use them in your system.

Javascript has already defined a semantic for `undefined` as variables that are not initialized or don't exist, and we can't change it. So we have no choice but to use `null` for explicit zero values.

> I never have to check if x === undefined || x === null.

I agree with that -- this happens when the two semantics are conflated, and we have a lot of this mess in Vexflow, because we don't have a clear convention. IMO, you should always know what to do with `null`, but you may not know what to do with `undefined`.

```typescript
if (x === null) {
  // defined predictable behavior, do something with it, or return recoverable error
}

if (x === undefined) {
  // unpredictable behaviour, can't continue, throw exception or return irrecoverable error
}

// this is bad, because it's conflating the two
if (x === null || x === undefined) { ... }
```

I think in general it's okay to use `undefined` when we don't know the zero value, and use `null` sparingly when we have a clear semantic for it. It makes a lot of sense to use `null` here.

Reply:
That's ok. However it does go into the differences:

>That said, there is one major difference - variables that aren't initialized (including function parameters where no argument was passed, among other things) are always undefined.

We are not using 'undefined' for unpredictible, can´t continue... only we use it as well for optional parameters in functions.

As I said, that's ok, as you said we should avoid the `if (x === null || x === undefined)`